cmd/clusterctl/client/yamlprocessor/simple_processor.go

```diff
@@ -51,21 +51,75 @@ func (tp *SimpleProcessor) GetTemplateName(_, flavor string) string {
 	return name
 }
 
-// GetVariables returns a list of the variables specified in the yaml.
-func (tp *SimpleProcessor) GetVariables(rawArtifact []byte) ([]string, error) {
+// Process processes the template blob of bytes and will return the final
+// yaml with values retrieved from the values getter.
+func (tp *SimpleProcessor) GetVariables(rawArtifact []byte) ([]VariableMetadata, error) {
 	strArtifact := convertLegacyVars(string(rawArtifact))
+	variables, err := getDefaultValues(strArtifact)
+	if err != nil {
+		return nil, err
+	}
+	res := make([]VariableMetadata, 0)
+	for _, v := range variables {
+		if v != nil {
+			res = append(res, v)
+		}
+	}
+	sort.Sort(ByName(res))
+	return res, nil
+}
 
-	variables, err := inspectVariables(strArtifact)
+func getDefaultValues(data string) (map[string]VariableMetadata, error) {
+	t, err := parse.Parse(data)
 	if err != nil {
 		return nil, err
 	}
+	seenMap := map[string]VariableMetadata{}
+	getAllDefault(t.Root, seenMap)
+	return seenMap, nil
+}
 
-	varNames := make([]string, 0, len(variables))
-	for k := range variables {
-		varNames = append(varNames, k)
+func getAllDefault(root parse.Node, seenMap map[string]VariableMetadata) {
+	st := ""
+	switch v := root.(type) {
+	case *parse.ListNode:
+		// iterate through the list node.
+		for _, ln := range v.Nodes {
+			getAllDefault(ln, seenMap)
+		}
+	case *parse.FuncNode:
+		if _, ok := seenMap[v.Param]; !ok {
+			if len(v.Args) > 0 {
+				for _, l := range v.Args {
+					st += getString(l)
+				}
+				seenMap[v.Param] = &variableMetadata{name: v.Param, required: false, defaultValue: st}
+			} else {
+				if len(v.Name) == 0 {
+					seenMap[v.Param] = &variableMetadata{name: v.Param, required: true, defaultValue: ""}
+				} else {
+					seenMap[v.Param] = &variableMetadata{name: v.Param, required: false, defaultValue: ""}
+				}
+			}
+		}
+	}
+}
+
+// from node generates default value string recursive function.
```

Comment:
this comment also needs fixing, it should start with the function name

https://golang.org/doc/effective_go#commentary